PermanentOrg/functional-test
# Signup Test Suite

This PR adds sign up, login, and cleanup test suites to the TypeScript functional test. This also creates the "Credentials" system, which saves logins between testing specs so that they can be reused between suites and cleaned up in the cleanup testing suite.

I tried to copy the algorithms used for generating usernames and passwords from the python test, so that is why the password generation code might look a bit weird (they were originally written in Python, but I decided to use the cryptographically secure random functions as well just to be safe).

Please leave any feedback on the `Credentials` object I created to handle created "sessions". I think eventually it should save both created *and* deleted sessions, so that suites know what the "auto increment" number should be when creating a new account after deletions, but I thought this was getting too into the weeds when that is an unusual (but not completely impossible) situation for running these tests. The module itself was designed under a lot of various compromises. Cypress tests run in a browser environment so Node.js specific code can't be run (such as `fs.readFileSync`) and in the `cypress.config.ts` file it's the opposite; the file is run in a Node.js specific domain so *browser* code won't run properly.

**Steps to test:**
1. `cd` into the `js` directory
2. Run `CYPRESS_BASE_URL=https://local.permanent.org npm run start`. You can change the BASE_URL to point to another environment.
3. Verify tests pass.
